src/elab.c

```diff
@@ -2046,15 +2046,18 @@ tree_t elab(tree_t top, jit_t *jit, cover_tagging_t *cover)
    case T_ENTITY:
       {
          tree_t arch = elab_pick_arch(tree_loc(top), top, &ctx);
+         tree_set_primary(e, arch);
```

Comment:
I think you can do this instead using the `T_HIER` object which is added to each `T_BLOCK` in the elaborated design as that has a reference to the design unit it was instantiated from, see my comment below.